src/PCE/PCE.jl

```diff
@@ -0,0 +1,244 @@
+extract monomials of the basis functions and the corresponding coeffiecients.
+
+# Returns
+`Vector` of `Dict`s mapping monomial of basis functions to its coefficient in the individual equations.
+"""
+function extract_basismonomial_coeffs(eqs::AbstractVector, pce::PCE)
+    basismonomial_coeffs = [extract_coeffs(eq, pce.sym_basis) for eq in eqs]
+    basismonomial_indices = []
+    for coeffs in basismonomial_coeffs
+        union!(basismonomial_indices,
+               [mono => get_basis_indices(mono) for mono in keys(coeffs)])
+    end
+    return basismonomial_coeffs, basismonomial_indices
+end
+
+# 3. compute inner products
+"""
+$(TYPEDSIGNATURES)
+
+Evaluate scalar products between all basis functions in `pce` and 
+basis monomials as characterized by `mono_indices`.
+"""
+function eval_scalar_products(mono_indices, pce::PCE)
+    uni_degs = [deg(op) for op in pce.tensor_basis.uni]
+    max_degs = uni_degs
+    for (mono, id) in mono_indices
+        max_degs = max.(max_degs, vec(sum(pce.tensor_basis.ind[id .+ 1, :], dims = 1)))
+    end
+    quad_deg = max.(ceil.(Int, 0.5 * (max_degs + uni_degs .+ 1)))
+
+    integrators = map((uni, deg) -> bump_degree(uni, deg), pce.tensor_basis.uni, quad_deg)
+    scalar_products = Dict()
+    for k in 1:dim(pce.tensor_basis)
+        scalar_products[k] = Dict(mono => computeSP(vcat(id, k - 1), pce.tensor_basis,
+                                                    integrators)
+                                  for (mono, id) in mono_indices)
+    end
+    return scalar_products
+end
+
+# 4. Galerkin projection
+"""
+$(TYPEDSIGNATURES)
+
+perform Galerkin projection of polynomial expressions characterized by `Dict`s mapping
+basis monomials to coefficients. 
+"""
+function galerkin_projection(bm_coeffs::Vector{<:Dict}, scalar_products::Dict,
+                             pce::PCE)
+    projected_eqs = []
+    scaling_factors = computeSP2(pce.tensor_basis)
+    for i in eachindex(bm_coeffs)
+        eqs = []
+        for k in 1:dim(pce.tensor_basis)
+            push!(eqs,
+                  sum(bm_coeffs[i][mono] * scalar_products[k][mono]
+                      for mono in keys(bm_coeffs[i])))
+        end
+        push!(projected_eqs, eqs ./ scaling_factors)
+    end
+    return projected_eqs
+end
+
+# 5. combine everything
+"""
+$(TYPEDSIGNATURES)
+
+perform Galerkin projection onto the `pce`.
+"""
+function pce_galerkin(eqs::AbstractVector, pce::PCE)
+    expanded_eqs = apply_ansatz(eqs, pce)
+    basismono_coeffs, basismono_idcs = extract_basismonomial_coeffs(expanded_eqs, pce)
+    scalar_products = eval_scalar_products(basismono_idcs, pce)
+    projected_eqs = galerkin_projection(basismono_coeffs, scalar_products, pce)
+    return projected_eqs
+end
+
+# 6. high-level interface
+# 6a. apply pce to explicit ODE
+"""
+$(TYPEDSIGNATURES)
+
+Generate moment equations of an `ODESystem` from a given `PCE`-Ansatz via Galerkin projection.
+"""
+function moment_equations(sys::ODESystem, pce::PCE)
```

Comment:
With this interface, users can set the PCE of their own choice, which is fantastic!

I think we can make it more powerful by doing this step for our users.

My idea is as follow

1. asking users to add the probability distributions of stochastic parameters as symbolic metadata. See [ModelingToolkit.jl - Symbolic metadata - Probability distributions](https://mtk.sciml.ai/dev/basics/Variable_metadata/#Probability-distributions).
2. we internally set the corresponding appropriate orthogonal polynomials according to the probability distribution.

I would be happy to adapt this PR if you think this is feasible.

Reply:
Sounds great to me. And at least for canonical distributions it not only should be feasible but even very easy to do. That said, I am unsure how easy it would be to support more obscure distributions. With the current implementation it is very easy. So for now, I would keep this and add the interface you are proposing for canonical distributions as well. There seems to be no harm in keeping both and they both address valid use cases.